Automattic/wp-calypso
# build: Merges desktop's dependency tree into the main dependency tree

Fixes #55184

## Background

WP Desktop (the electron app) lives in `./desktop`. However, it is not part of the main dependency tree. It has its own `yarn.lock` and it is not a yarn workspace. This is suboptimal for a number of reasons, like (but not limited to):

- It's invisible to Renovate updates or Dependabot security alerts
- Our CI doesn't check for package duplicates in `./desktop/yarn.lock`
- Many desktop commands require `yarn && cd desktop && yarn` [1]

## Changes

The main change is merging `./desktop/yarn.lock` with `./yarn.lock` and declare `./desktop` as a workspace. Unfortunately, there is no automated way to do this process so it required some manual work. This is the process I followed:

1) Delete `./desktop/.yarnrc` and declare it as a namespace in `./package.json`
2) Merge `./desktop/yarn.lock` into `./yarn.lock`
3) Resolve duplicated keys, trying to preserve `./yarn.lock` version when possible (so this doesn't accidentally bump a version in Calypso).

Still, both Calypso and Desktop will be getting new packages as part of this PR. There are two causes for this:

- As Desktop had some duplicated packages, those need to be de-duplicated. Example: `tar@npm:^6.0.2 can be deduped from tar@npm:6.1.0 to tar@npm:6.1.11`
- If the same version range was present in Calypso and Desktop and it resolved to a different version. For example, both have a dependency on `balanced-match@npm:^1.0.0`, but it was resolved as `balanced-match@1.0.2` in Desktop but `balanced-match@1.0.0` in Calypso.

I reviewed all differences in the dependency tree (as generated by `@automattic/effective-module-tree`) and all differences can be explained by one of those two cases. In any case, the differences are always transitive dependencies and small patch bumps, so it should be reasonably safe.

There are other minor changes in this PR worth mentioning: 

- Pin electron version to `12.0.3`. `electron-builder` needs to have a pinned version or an electron version installed in `./desktop/node_modules`. As we can't enforce the latter, I pinned the version to the same version we were already using.
- Change CI scripts so they don't do `yarn && cd desktop && yarn`

[1]: The downside is that dev flows that are not related to Desktop will now have to download Desktop packages, like `electron`. However, in my local laptop, this only adds a small delay: `yarn install` (measured by running `yarn distclean && time yarn`) takes `1:32.59` in this branch and `1:25.11` in `trunk`. 


## Testing instructions

* Verify Desktop E2E tests are still passing
* Smoke tests Desktop app.
* Checkout this branch, run `rm -fr **/node_modules && yarn` and then run this script:
    ```
    ( 
        setopt shwordsplit
        TARGET=dependency-tree
        FILES="./package.json ./client/package.json ./apps/editing-toolkit/package.json ./apps/notifications/package.json ./apps/o2-blocks/package.json ./apps/wpcom-block-editor/package.json ./test/e2e/package.json ./desktop/package.json"
        rm -f $TARGET && true
        for file in $FILES; do
            ./node_modules/.bin/effective-module-tree --root "$file" -o list >> $TARGET
        done
    )
    ```
    It should generate a `dependency-tree` file with the dependencies of all apps. Do the same with `trunk` and diff the resulting file. There will be some differences! Check them to make sure they are not big (i.e. small transitive dependencies and small patches/minor updates)
